### PR TITLE
Refactor yacfg Codebase and Update Project Dependencies

### PR DESCRIPTION
**Pull request by enkeys**
## Description
This pull request encompasses a wide range of refactoring updates for various yacfg modules aimed at improving code readability, maintainability, performance, and type safety. Updates include the addition of type annotations to function parameters and return types, simplification of code logic, reorganization of import statements, and improved exception handling. 

Here is an overview of the changes made:

- Dependencies in `pyproject.toml` have been updated, and the script entry for yacfg has been modified.
- Various modules such as `yacfg_cli`, `yacfg_batch_cli`, `cli_arguments`, `yacfg_batch`, `output`, `query`, `profiles`, `output`, `exceptions` have been significantly refactored for improved readability, maintainability, and performance.
- Improved handling of file paths, output files, and profile in `files.py`, `output.py`, `profiles.py` modules respectively.
- Changes to the metadata and descriptions for the `yacfg` and `yacfg-batch` packages.
- The `README.md` file and documentation have been updated with accurate information.
- The poetry config file has been updated and the version has been bumped to 0.10.0.
- `poetry.lock` file updated to ensure compatibility and stability of the project dependencies.

Each commit message has further details of the changes made.

### CODE REVIEW

File: src/yacfg/files.py
Comment (jiridanek) [inline src/yacfg/files.py:15]: I thought that `"""Get` (quotes and first word on a single line) is the generally preferred style. But OK, whatever. Let's turn on black in GHA some time next week anyways, or next month ;P

File: src/yacfg/files.py
Comment (jiridanek) [inline src/yacfg/files.py:19]: Duplicating the type annotation in a type comment is poor practice. I hope subsequent commits fix this!

File: src/yacfg/config_data.py
Comment (jiridanek) [inline src/yacfg/config_data.py:34]: `from __future__ import annotations` should be available in 3.7+, so that could be used, to avoid the quotes. (https://docs.python.org/3/library/__future__.html)

File: src/yacfg/yacfg.py
Comment (jiridanek) [inline src/yacfg/yacfg.py:256]: There's some compatibility package to allow using | instead of Union[] in older Pythons. I think that it should be adopted in the future, so we can have modern type annotations.

**Review by jiridanek — APPROVED** [review]
cant say i understood it all, but i did look at it and it seems legit

File: src/yacfg/config_data.py
Comment (enkeys) [reply src/yacfg/config_data.py:34]: Hey, you're absolutely right! I actually overlooked quite a few things. Thanks for being an extra set of eyes! I appreciate it!

File: src/yacfg/files.py
Comment (enkeys) [reply src/yacfg/files.py:19]: Yes, you're right. I should go through everything. I'm really excited about it. :grin:

Comment (jiridanek) [reply src/yacfg/files.py:19]: i wouldn't block this PR on it

File: src/yacfg/files.py
Comment (enkeys) [reply src/yacfg/files.py:15]: Indeed, I do find D213 and D212 in pydocstyle a bit challenging at the moment. By the way, I already utilize black for every commit. 😸

File: src/yacfg/yacfg.py
Comment (enkeys) [reply src/yacfg/yacfg.py:256]: Yeah, I'm facing the same situation. I should search for an alternative tool that can analyze all the docstrings. Moreover, since we have typing, we can consider removing this information from the docstrings altogether.

**Comment by enkeys** [comment]
> cant say i understood it all, but i did look at it and it seems legit

Thanks for checking out the stuff, If there's something you didn't understand or if you have any questions, just let me know. I'm happy to explain things better or give you more info. I've been refactoring and refactoring... It would deserve even more sincere attention, but we don't have the time to completely rewrite it from scratch.

**Comment by jiridanek** [comment]
> but we don't have the time to completely rewrite it from scratch.

Somebody (@gtully, et al) will rewrite this in Java anyways, to configure broker from properties, without init image that carres python...

Ideally, being able to configure fully a standalone instance (or cluster) conveniently, with single config file, and have the tool understand the schema defined in Java code would be nice... not sure if there's something to build it upon, though. I do not want to have to have the top level template in XML ;(

Maybe there would be some desire for ansible for this? I know that skupper resolved the Python 3.9 vulnerability on RHEL 8 by upgrading the image to RHEL 9, where Python 3.9 is patched (which IMO would've been sensible for broker too, but 3.8 there, that's already done deal). And skupper will have or has some ansible in upstream repo. So, that actually means it's back to jinja and python, because ansible ;P

File: src/yacfg/yacfg.py
Comment (jiridanek) [reply src/yacfg/yacfg.py:256]: Nobody reads the generated docs anyways, I think it's ok not to have types there for now. It's IDE where it matters.

Sphynx needed this, iirc, https://pypi.org/project/sphinx-autodoc-typehints/. That jupyter book thing I brought in should work with it somehow, https://jupyterbook.org/en/stable/explain/sphinx.html

If you want to leave that up to me, I can try looking, since I'm who introduced jupyter book in first place

**Pull request merged**
